app/ide-desktop/lib/dashboard/src/utilities/Debug.tsx

```diff
@@ -0,0 +1,60 @@
+/** @file Utilities related to debugging. */
+import * as React from 'react'
+
+import * as debugHooks from '#/hooks/debugHooks'
+
+/* eslint-disable no-restricted-properties */
+
+// =============
+// === Debug ===
+// =============
+
+/** Props for a {@Link Debug}. */
+interface DebugProps {
+  readonly monitorProps?: boolean
+  readonly monitorRender?: boolean
+  readonly children: JSX.Element
+}
+
+/** A component that adds debugging info to its direct child. */
+export default function Debug(props: DebugProps) {
+  const { monitorProps = false, monitorRender = false, children } = props
+  const childPropsRaw: unknown = children.props
+  const childProps: object = typeof childPropsRaw === 'object' ? childPropsRaw ?? {} : {}
+  const propsValues: unknown[] = Object.values(childProps)
+  const typeRaw: unknown = children.type
+  const typeName = typeof typeRaw === 'function' ? typeRaw.name : String(children.type)
+  debugHooks.useMonitorDependencies(
+    [children.key, ...propsValues],
+    typeName,
+    ['key', ...Object.keys(childProps)],
+    monitorProps
+  )
+
+  const patchedChildProps = Object.fromEntries(
+    Object.entries(childProps).map(([key, value]: [string, unknown]) => [
+      key,
+      typeof value !== 'function'
+        ? value
+        : (...args: unknown[]) => {
+            console.group(`[Debug(${typeName})] Prop '${key}' called with args: [`, ...args, ']')
+            const result: unknown = value(...args)
+            console.log('Returned', result)
+            console.groupEnd()
+            return result
+          },
+    ])
+  )
+
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+  const Component = children.type
+  if (monitorRender) {
+    console.group(`[Debug(${typeName})] Rendering`, Component, patchedChildProps)
+  }
+  const element = <Component {...patchedChildProps} />
+  if (monitorRender) {
```

Comment:
Render doesn't finish after calling `<Component {...patchedChildProps} />`

Reply:
oops, true... i'm not sure whether it's useful to track renders, so i'm thinking of just renaming it from `Rendering` to `Constructing React Node` or similar. thoughts?

Reply:
i guess i'll test with nesting `Debug` and do whatever makes the most sense with that

Reply:
should be fixed now?

Reply:
> Constructing React Node or similar. thoughts

For me, it makes no sense... constructing a node is a trivial operation and gives no debug info at all.